plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java

```diff
@@ -736,26 +737,54 @@ public synchronized void commentTable(String databaseName, String tableName, Opt
     @Override
     public synchronized void commentColumn(String databaseName, String tableName, String columnName, Optional<String> comment)
     {
-        alterTable(databaseName, tableName, oldTable -> {
-            if (oldTable.getColumn(columnName).isEmpty()) {
-                SchemaTableName name = new SchemaTableName(databaseName, tableName);
-                throw new ColumnNotFoundException(name, columnName);
-            }
+        alterTable(databaseName, tableName, table -> {
+            List<Column> dataColumns = table.getDataColumns();
+            List<Column> partitionColumns = table.getPartitionColumns();
 
-            ImmutableList.Builder<Column> newDataColumns = ImmutableList.builder();
-            for (Column fieldSchema : oldTable.getDataColumns()) {
-                if (fieldSchema.getName().equals(columnName)) {
-                    newDataColumns.add(new Column(columnName, fieldSchema.getType(), comment, fieldSchema.getProperties()));
-                }
-                else {
-                    newDataColumns.add(fieldSchema);
-                }
+            Optional<Integer> matchingDataColumn = indexOfColumnWithName(dataColumns, columnName);
+            Optional<Integer> matchingPartitionColumn = indexOfColumnWithName(partitionColumns, columnName);
+
+            if (matchingDataColumn.isPresent() && matchingPartitionColumn.isPresent()) {
+                throw new TrinoException(HIVE_INVALID_METADATA, "Found two columns with names matching " + columnName);
+            }
+            if (matchingDataColumn.isEmpty() && matchingPartitionColumn.isEmpty()) {
+                throw new ColumnNotFoundException(new SchemaTableName(databaseName, tableName), columnName);
             }
 
-            return oldTable.withDataColumns(currentVersion, newDataColumns.build());
+            return table
+                    .withDataColumns(currentVersion, matchingDataColumn.map(index -> setColumnCommentForIndex(dataColumns, index, comment)).orElse(dataColumns))
+                    .withPartitionColumns(currentVersion, matchingPartitionColumn.map(index -> setColumnCommentForIndex(partitionColumns, index, comment)).orElse(partitionColumns));
         });
     }
 
+    private static Optional<Integer> indexOfColumnWithName(List<Column> columns, String columnName)
+    {
+        Optional<Integer> index = Optional.empty();
+        for (int i = 0; i < columns.size(); i++) {
+            if (columns.get(i).getName().equals(columnName)) {
+                index.ifPresent(ignored -> {
```

Comment:
when can this situation occur?
is it worth to handle it?

Reply:
I don't know. I've copied it from Glue meta store. Better safe than sorry.